src/super_gradients/training/losses/ssd_loss.py

```diff
@@ -1,66 +1,129 @@
-        destination box which overlaps it over 0.5 (IoU). so some gt bboxes can be duplicated to a few destination boxes
-        :param targets: a tensor containing the boxes for a single image. shape [num_boxes, 5] (x,y,w,h,label)
-        :return: two tensors
-            boxes - shape of dboxes [4, num_dboxes] (x,y,w,h)
-            labels - sahpe [num_dboxes]
+        creates tensors with target boxes and labels for each dboxes, so with the same len as dboxes.
+
+        * Each GT is assigned with a grid cell with the highest IoU, this creates a pair for each GT and some cells;
+        * The rest of grid cells are assigned to a GT with the highest IoU, assuming it's > self.iou_thresh;
+          If this condition is not met the grid cell is marked as background
+
+        GT-wise: one to many
+        Grid-cell-wise: one to one
+
+        :param targets: a tensor containing the boxes for a single image;
+                        shape [num_boxes, 6] (image_id, label, x, y, w, h)
+        :return:        two tensors
+                        boxes - shape of dboxes [4, num_dboxes] (x,y,w,h)
+                        labels - sahpe [num_dboxes]
         """
-        target_locations = self.dboxes.data.clone().squeeze()
-        target_labels = torch.zeros((self.dboxes.data.shape[2])).to(self.dboxes.device)
+        device = targets.device
+        each_cell_target_locations = self.dboxes.data.clone().squeeze()
+        each_cell_target_labels = torch.zeros((self.dboxes.data.shape[2])).to(device)
 
         if len(targets) > 0:
-            boxes = targets[:, 2:]
-            ious = calculate_bbox_iou_matrix(boxes, self.dboxes.data.squeeze().T, x1y1x2y2=False)
+            target_boxes = targets[:, 2:]
+            target_labels = targets[:, 1]
+            ious = calculate_bbox_iou_matrix(target_boxes, self.dboxes.data.squeeze().T, x1y1x2y2=False)
+
+            # one best GT for EACH cell (does not guarantee that all GTs will be used)
+            best_target_per_cell, best_target_per_cell_index = ious.max(0)
 
-            values, indices = torch.max(ious, dim=0)
-            mask = values > 0.5
+            # one best grid cell for EACH target
+            best_cell_per_target, best_cell_per_target_index = ious.max(1)
+            # make sure EACH target has a grid cell assigned
+            best_target_per_cell_index[best_cell_per_target_index] = torch.arange(len(targets)).to(device)
+            best_target_per_cell[best_cell_per_target_index] = 2.
 
-            target_locations[:, mask] = targets[indices[mask], 2:].T
-            target_labels[mask] = targets[indices[mask], 1]
+            mask = best_target_per_cell > self.iou_thresh
+            each_cell_target_locations[:, mask] = target_boxes[best_target_per_cell_index[mask]].T
+            each_cell_target_labels[mask] = target_labels[best_target_per_cell_index[mask]] + 1
```

Comment:
Where are we taking care of this when computing mAP ?

Reply:
I'm still unhappy about using offset, because it requires taking care of the shift in the post-prediction callback, any other external code like COCO API where ids are mapped to classes, implementing offset in any new potential dataset. And adding more code = more things can break. While this tiny `+1` a few lines below does the trick. 
LMK if you still want me to add the offset 🙃

Reply:
Not really, NMS accepts objectness and other classes, then during execution it doesn't treat objectness as a class, as expected, and returns the class id starting from 0 (ignoring objectness). So whenever indices are involved, they are 0-based, accept for the target constructed in the loss here

Reply:
I know, all Im saying is to let the users be aware that once we feed class labels to SSD post prediction callback, they should beware that the returned class ids (**for predictions**) will be 0 based. Just a small note in docs thats all.